File: native/android/VueNativeCore/src/main/kotlin/com/vuenative/core/Modules/AudioModule.kt
package com.vuenative.core

import android.content.Context
import android.media.MediaPlayer
import android.media.MediaRecorder
import android.os.Build
import android.os.Handler
import android.os.Looper
import java.io.File
import java.util.UUID

/**
 * AudioModule — backs the useAudio() composable.
 *
 * Playback: MediaPlayer (play, pause, resume, stop, seek, setVolume)
 * Recording: MediaRecorder (startRecording, stopRecording, pauseRecording, resumeRecording)
 * Events: audio:progress, audio:complete, audio:error
 */
class AudioModule : NativeModule {
    override val moduleName = "Audio"

    private var player: MediaPlayer? = null
    private var recorder: MediaRecorder? = null
    private var recordingPath: String? = null
    private var bridge: NativeBridge? = null
    private var context: Context? = null
    private var isPlaying = false

    private val handler = Handler(Looper.getMainLooper())
    private var progressRunnable: Runnable? = null

    override fun initialize(context: Context, bridge: NativeBridge) {
        this.context = context
        this.bridge = bridge
    }

    override fun invoke(
        method: String,
        args: List<Any?>,
        bridge: NativeBridge,
        callback: (Any?, String?) -> Unit
    ) {
        when (method) {
            "play" -> {
                val uri = args.getOrNull(0)?.toString() ?: ""
                @Suppress("UNCHECKED_CAST")
                val options = (args.getOrNull(1) as? Map<String, Any?>) ?: emptyMap()
                play(uri, options, callback)
            }
            "pause" -> pause(callback)
            "resume" -> resume(callback)
            "stop" -> stop(callback)
            "seek" -> {
                val position = when (val v = args.getOrNull(0)) {
                    is Number -> v.toDouble()
                    else -> 0.0
                }
                seek(position, callback)
            }
            "setVolume" -> {
                val volume = when (val v = args.getOrNull(0)) {
                    is Number -> v.toFloat()
                    else -> 1.0f
                }
                setVolume(volume, callback)
            }
            "startRecording" -> {
                @Suppress("UNCHECKED_CAST")
                val options = (args.getOrNull(0) as? Map<String, Any?>) ?: emptyMap()
                startRecording(options, callback)
            }
            "stopRecording" -> stopRecording(callback)
            "pauseRecording" -> pauseRecording(callback)
            "resumeRecording" -> resumeRecording(callback)
            "getStatus" -> getStatus(callback)
            else -> callback(null, "AudioModule: Unknown method '$method'")
        }
    }

    // MARK: - Playback

    private fun play(uri: String, options: Map<String, Any?>, callback: (Any?, String?) -> Unit) {
        handler.post {
            try {
                stopProgressReporting()
                player?.release()
                player = null

                val mp = MediaPlayer()
                mp.setDataSource(uri)

                val volume = (options["volume"] as? Number)?.toFloat() ?: 1.0f
                val loop = options["loop"] as? Boolean ?: false

                mp.setVolume(volume, volume)
                mp.isLooping = loop

                mp.setOnPreparedListener {
                    it.start()
                    isPlaying = true
                    startProgressReporting()
                    callback(mapOf(
                        "duration" to it.duration.toDouble() / 1000.0,
                        "currentTime" to 0.0
                    ), null)
                }

                mp.setOnCompletionListener {
                    isPlaying = false
                    stopProgressReporting()
                    bridge?.dispatchGlobalEvent("audio:complete", emptyMap())
                }

                mp.setOnErrorListener { _, what, extra ->
                    isPlaying = false
                    stopProgressReporting()
                    val msg = "MediaPlayer error: what=$what extra=$extra"
                    bridge?.dispatchGlobalEvent("audio:error", mapOf("message" to msg))
                    true
                }

                player = mp
                mp.prepareAsync()
            } catch (e: Exception) {
                callback(null, "Failed to play audio: ${e.message}")
            }
        }
    }

    private fun pause(callback: (Any?, String?) -> Unit) {
        handler.post {
            player?.pause()
            isPlaying = false
            stopProgressReporting()
            callback(null, null)
        }
    }

    private fun resume(callback: (Any?, String?) -> Unit) {
        handler.post {
            player?.start()
            isPlaying = true
            startProgressReporting()
            callback(null, null)
        }
    }

    private fun stop(callback: (Any?, String?) -> Unit) {
        handler.post {
            stopProgressReporting()
            player?.release()
            player = null
            isPlaying = false
            callback(null, null)
        }
    }

    private fun seek(position: Double, callback: (Any?, String?) -> Unit) {
        handler.post {
            val ms = (position * 1000).toInt()
            player?.seekTo(ms)
            callback(null, null)
        }
    }

    private fun setVolume(volume: Float, callback: (Any?, String?) -> Unit) {
        handler.post {
            val v = volume.coerceIn(0f, 1f)
            player?.setVolume(v, v)
            callback(null, null)
        }
    }

    // MARK: - Progress Reporting

    private fun startProgressReporting() {
        stopProgressReporting()
        val runnable = object : Runnable {
            override fun run() {
                val mp = player ?: return
                if (isPlaying) {
                    bridge?.dispatchGlobalEvent("audio:progress", mapOf(
                        "currentTime" to mp.currentPosition.toDouble() / 1000.0,
                        "duration" to mp.duration.toDouble() / 1000.0
                    ))
                    handler.postDelayed(this, 250)
                }
            }
        }
        progressRunnable = runnable
        handler.postDelayed(runnable, 250)
    }

    private fun stopProgressReporting() {
        progressRunnable?.let { handler.removeCallbacks(it) }
        progressRunnable = null
    }

    // MARK: - Recording

    private fun startRecording(options: Map<String, Any?>, callback: (Any?, String?) -> Unit) {
        handler.post {
            try {
                val ctx = context ?: run {
                    callback(null, "No context available")
                    return@post
                }

                val quality = options["quality"]?.toString() ?: "medium"
                val format = options["format"]?.toString() ?: "m4a"

                val ext = if (format == "wav") "wav" else "m4a"
                val file = File(ctx.cacheDir, "${UUID.randomUUID()}.$ext")
                recordingPath = file.absolutePath

                val mr = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                    MediaRecorder(ctx)
                } else {
                    @Suppress("DEPRECATION")
                    MediaRecorder()
                }

                mr.setAudioSource(MediaRecorder.AudioSource.MIC)

                if (format == "wav") {
                    // WAV not directly supported by MediaRecorder — use 3GPP as fallback
                    mr.setOutputFormat(MediaRecorder.OutputFormat.THREE_GPP)
                    mr.setAudioEncoder(MediaRecorder.AudioEncoder.AMR_NB)
                } else {
                    mr.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
                    mr.setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
                    when (quality) {
                        "low" -> {
                            mr.setAudioSamplingRate(22050)
                            mr.setAudioEncodingBitRate(32000)
                        }
                        "high" -> {
                            mr.setAudioSamplingRate(44100)
                            mr.setAudioEncodingBitRate(128000)
                        }
                        else -> {
                            mr.setAudioSamplingRate(44100)
                            mr.setAudioEncodingBitRate(64000)
                        }
                    }
                }

                mr.setOutputFile(file.absolutePath)
                mr.prepare()
                mr.start()
                recorder = mr

                callback(mapOf("uri" to "file://${file.absolutePath}"), null)
            } catch (e: Exception) {
                callback(null, "Failed to start recording: ${e.message}")
            }
        }
    }

    private fun stopRecording(callback: (Any?, String?) -> Unit) {
        handler.post {
            val mr = recorder
            val path = recordingPath
            if (mr == null || path == null) {
                callback(null, "No active recording")
                return@post
            }
            try {
                mr.stop()
                mr.release()
                recorder = null

                // Get duration by briefly opening the file with MediaPlayer
                val mp = MediaPlayer()
                mp.setDataSource(path)
                mp.prepare()
                val duration = mp.duration.toDouble() / 1000.0
                mp.release()

                callback(mapOf(
                    "uri" to "file://$path",
                    "duration" to duration
                ), null)
            } catch (e: Exception) {
                callback(null, "Failed to stop recording: ${e.message}")
            }
        }
    }

    private fun pauseRecording(callback: (Any?, String?) -> Unit) {
        handler.post {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                recorder?.pause()
            }
            callback(null, null)
        }
    }

    private fun resumeRecording(callback: (Any?, String?) -> Unit) {
        handler.post {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                recorder?.resume()
            }
            callback(null, null)
        }
    }

    // MARK: - Status

    private fun getStatus(callback: (Any?, String?) -> Unit) {
        handler.post {
            val status = mutableMapOf<String, Any>(
                "isPlaying" to isPlaying,
                "isRecording" to (recorder != null)
            )
            player?.let {
                status["currentTime"] = it.currentPosition.toDouble() / 1000.0
                status["duration"] = it.duration.toDouble() / 1000.0
            }
            callback(status, null)
        }
    }

    override fun destroy() {
        stopProgressReporting()
        player?.release()
        player = null
        recorder?.release()
        recorder = null
    }
}

File: native/ios/VueNativeCore/Sources/VueNativeCore/Modules/AudioModule.swift
#if canImport(UIKit)
import UIKit
import AVFoundation

/// Native module for audio playback and recording.
///
/// Methods:
///   - play(uri: String, options?: Object)  -- play audio from URI
///   - pause()                              -- pause playback
///   - resume()                             -- resume playback
///   - stop()                               -- stop playback and release player
///   - seek(position: Number)               -- seek to position in seconds
///   - setVolume(volume: Number)            -- set volume 0.0–1.0
///   - startRecording(options?: Object)     -- start audio recording
///   - stopRecording()                      -- stop recording, returns { uri, duration }
///   - pauseRecording()                     -- pause recording
///   - resumeRecording()                    -- resume recording
///   - getStatus()                          -- returns current playback status
///
/// Events (via bridge.dispatchGlobalEvent):
///   - audio:progress { currentTime, duration }
///   - audio:complete {}
///   - audio:error { message }
final class AudioModule: NativeModule {
    let moduleName = "Audio"

    private var player: AVAudioPlayer?
    private var recorder: AVAudioRecorder?
    private var displayLink: CADisplayLink?
    private var isPlaying = false
    private var bridge: NativeBridge? { NativeBridge.shared }

    // MARK: - Delegate to forward completion events
    private var playerDelegate: AudioPlayerDelegate?

    func invoke(method: String, args: [Any], callback: @escaping (Any?, String?) -> Void) {
        switch method {
        case "play":
            let uri = args.first as? String ?? ""
            let options = (args.count > 1 ? args[1] as? [String: Any] : nil) ?? [:]
            play(uri: uri, options: options, callback: callback)

        case "pause":
            pause(callback: callback)

        case "resume":
            resume(callback: callback)

        case "stop":
            stop(callback: callback)

        case "seek":
            let position = (args.first as? Double) ?? (args.first as? Int).map(Double.init) ?? 0
            seek(position: position, callback: callback)

        case "setVolume":
            let volume = (args.first as? Double) ?? (args.first as? Int).map(Double.init) ?? 1.0
            setVolume(Float(volume), callback: callback)

        case "startRecording":
            let options = args.first as? [String: Any] ?? [:]
            startRecording(options: options, callback: callback)

        case "stopRecording":
            stopRecording(callback: callback)

        case "pauseRecording":
            pauseRecording(callback: callback)

        case "resumeRecording":
            resumeRecording(callback: callback)

        case "getStatus":
            getStatus(callback: callback)

        default:
            callback(nil, "AudioModule: Unknown method '\(method)'")
        }
    }

    // MARK: - Playback

    private func play(uri: String, options: [String: Any], callback: @escaping (Any?, String?) -> Void) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            // Stop any existing playback
            self.stopProgressReporting()
            self.player?.stop()

            do {
                try AVAudioSession.sharedInstance().setCategory(.playback, mode: .default)
                try AVAudioSession.sharedInstance().setActive(true)
            } catch {
                callback(nil, "Failed to set audio session: \(error.localizedDescription)")
                return
            }

            guard let url = URL(string: uri) else {
                callback(nil, "Invalid audio URI: \(uri)")
                return
            }

            // Check if it's a remote URL — download first
            if url.scheme == "http" || url.scheme == "https" {
                self.downloadAndPlay(url: url, options: options, callback: callback)
            } else {
                self.playLocal(url: url, options: options, callback: callback)
            }
        }
    }

    private func downloadAndPlay(url: URL, options: [String: Any], callback: @escaping (Any?, String?) -> Void) {
        URLSession.shared.dataTask(with: url) { [weak self] data, _, error in
            DispatchQueue.main.async {
                guard let self = self else { return }
                if let error = error {
                    callback(nil, "Failed to download audio: \(error.localizedDescription)")
                    self.bridge?.dispatchGlobalEvent("audio:error", payload: ["message": error.localizedDescription])
                    return
                }
                guard let data = data else {
                    callback(nil, "No audio data received")
                    return
                }
                do {
                    let player = try AVAudioPlayer(data: data)
                    self.setupPlayer(player, options: options, callback: callback)
                } catch {
                    callback(nil, "Failed to initialize player: \(error.localizedDescription)")
                }
            }
        }.resume()
    }

    private func playLocal(url: URL, options: [String: Any], callback: @escaping (Any?, String?) -> Void) {
        do {
            let player = try AVAudioPlayer(contentsOf: url)
            setupPlayer(player, options: options, callback: callback)
        } catch {
            callback(nil, "Failed to play audio: \(error.localizedDescription)")
        }
    }

    private func setupPlayer(_ player: AVAudioPlayer, options: [String: Any], callback: @escaping (Any?, String?) -> Void) {
        let volume = Float(options["volume"] as? Double ?? 1.0)
        let loop = options["loop"] as? Bool ?? false

        player.volume = volume
        player.numberOfLoops = loop ? -1 : 0

        let delegate = AudioPlayerDelegate { [weak self] successfully in
            guard let self = self else { return }
            self.isPlaying = false
            self.stopProgressReporting()
            self.bridge?.dispatchGlobalEvent("audio:complete", payload: [:])
        }
        player.delegate = delegate
        self.playerDelegate = delegate
        self.player = player

        player.prepareToPlay()
        player.play()
        self.isPlaying = true
        self.startProgressReporting()

        callback([
            "duration": player.duration,
            "currentTime": 0.0,
        ], nil)
    }

    private func pause(callback: @escaping (Any?, String?) -> Void) {
        DispatchQueue.main.async { [weak self] in
            self?.player?.pause()
            self?.isPlaying = false
            self?.stopProgressReporting()
            callback(nil, nil)
        }
    }

    private func resume(callback: @escaping (Any?, String?) -> Void) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.player?.play()
            self.isPlaying = true
            self.startProgressReporting()
            callback(nil, nil)
        }
    }

    private func stop(callback: @escaping (Any?, String?) -> Void) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.player?.stop()
            self.player = nil
            self.playerDelegate = nil
            self.isPlaying = false
            self.stopProgressReporting()
            callback(nil, nil)
        }
    }

    private func seek(position: Double, callback: @escaping (Any?, String?) -> Void) {
        DispatchQueue.main.async { [weak self] in
            self?.player?.currentTime = position
            callback(nil, nil)
        }
    }

    private func setVolume(_ volume: Float, callback: @escaping (Any?, String?) -> Void) {
        DispatchQueue.main.async { [weak self] in
            self?.player?.volume = max(0, min(1, volume))
            callback(nil, nil)
        }
    }

    // MARK: - Progress Reporting

    private func startProgressReporting() {
        stopProgressReporting()
        let link = CADisplayLink(target: self, selector: #selector(reportProgress))
        // Report at ~4 Hz (every 15 frames at 60fps)
        link.preferredFrameRateRange = CAFrameRateRange(minimum: 4, maximum: 4, preferred: 4)
        link.add(to: .main, forMode: .common)
        displayLink = link
    }

    private func stopProgressReporting() {
        displayLink?.invalidate()
        displayLink = nil
    }

    @objc private func reportProgress() {
        guard let player = player, isPlaying else { return }
        bridge?.dispatchGlobalEvent("audio:progress", payload: [
            "currentTime": player.currentTime,
            "duration": player.duration,
        ])
    }

    // MARK: - Recording

    private func startRecording(options: [String: Any], callback: @escaping (Any?, String?) -> Void) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            do {
                try AVAudioSession.sharedInstance().setCategory(.playAndRecord, mode: .default)
                try AVAudioSession.sharedInstance().setActive(true)
            } catch {
                callback(nil, "Failed to set audio session for recording: \(error.localizedDescription)")
                return
            }

            let quality = options["quality"] as? String ?? "medium"
            let format = options["format"] as? String ?? "m4a"

            let ext = format == "wav" ? "wav" : "m4a"
            let url = FileManager.default.temporaryDirectory
                .appendingPathComponent(UUID().uuidString + ".\(ext)")

            var settings: [String: Any] = [:]
            if format == "wav" {
                settings = [
                    AVFormatIDKey: Int(kAudioFormatLinearPCM),
                    AVSampleRateKey: quality == "high" ? 44100.0 : 22050.0,
                    AVNumberOfChannelsKey: 1,
                    AVLinearPCMBitDepthKey: 16,
                    AVLinearPCMIsFloatKey: false,
                ]
            } else {
                let sampleRate: Double
                let bitRate: Int
                switch quality {
                case "low":  sampleRate = 22050; bitRate = 32000
                case "high": sampleRate = 44100; bitRate = 128000
                default:     sampleRate = 44100; bitRate = 64000
                }
                settings = [
                    AVFormatIDKey: Int(kAudioFormatMPEG4AAC),
                    AVSampleRateKey: sampleRate,
                    AVNumberOfChannelsKey: 1,
                    AVEncoderAudioQualityKey: AVAudioQuality.medium.rawValue,
                    AVEncoderBitRateKey: bitRate,
                ]
            }

            do {
                let recorder = try AVAudioRecorder(url: url, settings: settings)
                recorder.prepareToRecord()
                recorder.record()
                self.recorder = recorder
                callback(["uri": url.absoluteString], nil)
            } catch {
                callback(nil, "Failed to start recording: \(error.localizedDescription)")
            }
        }
    }

    private func stopRecording(callback: @escaping (Any?, String?) -> Void) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self, let recorder = self.recorder else {
                callback(nil, "No active recording")
                return
            }
            let duration = recorder.currentTime
            let uri = recorder.url.absoluteString
            recorder.stop()
            self.recorder = nil
            callback(["uri": uri, "duration": duration], nil)
        }
    }

    private func pauseRecording(callback: @escaping (Any?, String?) -> Void) {
        DispatchQueue.main.async { [weak self] in
            self?.recorder?.pause()
            callback(nil, nil)
        }
    }

    private func resumeRecording(callback: @escaping (Any?, String?) -> Void) {
        DispatchQueue.main.async { [weak self] in
            self?.recorder?.record()
            callback(nil, nil)
        }
    }

    // MARK: - Status

    private func getStatus(callback: @escaping (Any?, String?) -> Void) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { callback(nil, nil); return }
            var status: [String: Any] = [
                "isPlaying": self.isPlaying,
                "isRecording": self.recorder?.isRecording ?? false,
            ]
            if let player = self.player {
                status["currentTime"] = player.currentTime
                status["duration"] = player.duration
                status["volume"] = player.volume
            }
            callback(status, nil)
        }
    }
}

// MARK: - AVAudioPlayerDelegate wrapper

private final class AudioPlayerDelegate: NSObject, AVAudioPlayerDelegate {
    private let onComplete: (Bool) -> Void

    init(onComplete: @escaping (Bool) -> Void) {
        self.onComplete = onComplete
    }

    func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
        onComplete(flag)
    }

    func audioPlayerDecodeErrorDidOccur(_ player: AVAudioPlayer, error: Error?) {
        // Treat decode error as completion failure
        onComplete(false)
    }
}
#endif
